### PR TITLE
Inherit MSIDAadAuthorityCache from NSCache

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m [hunk 1]
```diff
@@ -74,22 +53,19 @@ + (MSIDAadAuthorityCache *)sharedInstance
...
-    
-    [self getWriteLock];
-    BOOL ret = [self processImpl:metadata authority:authority openIdConfigEndpoint:openIdConfigEndpoint context:context error:error];
-    pthread_rwlock_unlock(&_rwLock);
-    
-    return ret;
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        NSError *error;
+        BOOL result = [self processImpl:metadata authority:authority openIdConfigEndpoint:openIdConfigEndpoint context:context error:&error];
+        dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (unpluggedk) [inline IdentityCore/src/validation/MSIDAadAuthorityCache.m]: Why does this need to be in a main queue?

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m [hunk 2]
```diff
@@ -74,22 +53,19 @@ + (MSIDAadAuthorityCache *)sharedInstance
...
-    {
-        CHECK_CLASS_TYPE(metadata, NSArray, @"JSON metadata from authority validation is not an array");
-    }
-    
-    [self getWriteLock];
-    BOOL ret = [self processImpl:metadata authority:authority openIdConfigEndpoint:openIdConfigEndpoint context:context error:error];
-    pthread_rwlock_unlock(&_rwLock);
-    
-    return ret;
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```
Comment (unpluggedk) [inline IdentityCore/src/validation/MSIDAadAuthorityCache.m]: same as above, isn't this entire function synchronous?

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m [hunk 3]
```diff
@@ -74,22 +53,19 @@ + (MSIDAadAuthorityCache *)sharedInstance
     return singleton;
 }
 
-- (BOOL)processMetadata:(NSArray<NSDictionary *> *)metadata
+- (void)processMetadata:(NSArray<NSDictionary *> *)metadata
```
Comment (unpluggedk) [inline IdentityCore/src/validation/MSIDAadAuthorityCache.m]: This makes the method look asynchronous, isn't this a synchronous method?

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAadAuthorityCache.m]: Json parsing is heavy operation. If we get huge json file with metadata, parsing on a main queue  could block it for a while. That is why we do json parsing on a background thread and then return completion block on a main queue.

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAadAuthorityCache.m]: This method is async.

File: IdentityCore/src/MSIDCache.m [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        object = [self.container objectForKey:key];
+    });
+    
+    return object;
+}
+
+- (void)setObject:(id)obj forKey:(id)key
+{
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        [self.container setObject:obj forKey:key];
```
Comment (oldalton) [inline IdentityCore/src/MSIDCache.m]: nit: should we use setValue:forKey: instead to handle nil cases or subscripts?

File: IdentityCore/tests/MSIDCacheTests.m [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    __auto_type cache = [MSIDCache new];
+    
+    [cache setObject:@"v1" forKey:@"k1"];
+    id object = [cache objectForKey:@"k1"];
+    
+    XCTAssertEqual([cache count], 1);
+    XCTAssertEqualObjects(object, @"v1");
+}
+
+- (void)testOnRemoveObjectForKey_shouldRemoveObject
```
Comment (oldalton) [inline IdentityCore/tests/MSIDCacheTests.m]: nit: please follow our agreed unit test style of testMethod_whenCondition_shouldResult

Comment (antrix1989) [reply IdentityCore/tests/MSIDCacheTests.m]: done

File: IdentityCore/src/MSIDCache.m [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+@property (nonatomic) NSMutableDictionary *container;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
+
+@implementation MSIDCache
+
+- (instancetype)init
+{
+    if (!(self = [super init]))
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/MSIDCache.m]: done

**Pull request merged**